source/rofi-icon-fetcher.c

```diff
@@ -91,6 +100,150 @@ typedef struct {
 
+static void rofi_icon_fetcher_load_thumbnailers(const gchar *path) {
+  gchar *thumb_path = g_build_filename(path, "thumbnailers", NULL);
+
+  GDir *dir = g_dir_open(thumb_path, 0, NULL);
+  
+  if (!dir) {
+    g_free(thumb_path);
+    return;
+  }
+
+  const gchar *dirent;
+  
+  while ((dirent = g_dir_read_name(dir))) {
+    if (!g_str_has_suffix(dirent, THUMBNAILER_EXTENSION))
+      continue;
+
+    gchar *filename = g_build_filename(thumb_path, dirent, NULL);
+    GKeyFile *key_file = g_key_file_new();
+    GError *error = NULL;
+
+    if (!g_key_file_load_from_file(key_file, filename, 0, &error)) {
+      g_warning("Error loading thumbnailer %s: %s", filename, error->message);
+      g_error_free(error);
+    } else {
+      gchar *command = g_key_file_get_string(
+        key_file, THUMBNAILER_ENTRY_GROUP, "Exec", NULL);
+      gchar **mime_types = g_key_file_get_string_list(
+        key_file, THUMBNAILER_ENTRY_GROUP, "MimeType", NULL, NULL);
+
+      if (mime_types && command) {
+        guint i;
+        for (i = 0; mime_types[i] != NULL; i++) {
+          if (!g_hash_table_lookup(
+              rofi_icon_fetcher_data->thumbnailers, mime_types[i])) {
+            g_info("Loading thumbnailer %s for mimetype %s", filename, mime_types[i]);
+            g_hash_table_insert(
+              rofi_icon_fetcher_data->thumbnailers,
+              g_strdup(mime_types[i]),
+              g_strdup(command)
+            );
+          }
+        }
+      }
+
+      if (mime_types) g_strfreev(mime_types);
+      if (command) g_free(command);
+    }
+
+    g_key_file_free(key_file);
+    g_free(filename);
+  }
+
+  g_dir_close(dir);
+  g_free(thumb_path);
+}
+
+static gchar** setup_thumbnailer_command(const gchar *command,
+                                         const gchar *filename,
+                                         const gchar *encoded_uri,
+                                         const gchar *output_path,
+                                         int size) {
+  gchar **command_parts = g_strsplit(command, " ", 0);
+  gchar **command_args = NULL;
+  
+  GStrvBuilder *cmd_builder = g_strv_builder_new();
+  
+  if (command_parts) {
+    // set process niceness value to 19 (low priority)
+    g_strv_builder_add(cmd_builder, "nice");
+    g_strv_builder_add(cmd_builder, "-n");
+    g_strv_builder_add(cmd_builder, "19");
+    
+    // add executable and arguments of the thumbnailer to the list
+    guint i;
+    for (i = 0; command_parts[i] != NULL; i++) {
+      if (strcmp(command_parts[i], "%i") == 0) {
+        g_strv_builder_add(cmd_builder, filename);
+      } else if (strcmp(command_parts[i], "%u") == 0) {
+        g_strv_builder_add(cmd_builder, encoded_uri);
+      } else if (strcmp(command_parts[i], "%o") == 0) {
+        g_strv_builder_add(cmd_builder, output_path);
+      } else if (strcmp(command_parts[i], "%s") == 0) {
+        char size_str[33];
+        sprintf(size_str, "%d", size);
```

Comment:
To stop static code analysis tools, lets use snprintf.

Reply:
ok this one was easy :)